### PR TITLE
[DO NOT SQUASH] Support nn-Meter in Retiarii framework

### PR DESCRIPTION
**Pull request by ultmaster**

### CODE REVIEW

File: nni/retiarii/converter/graph_gen.py [hunk 1]
```diff
@@ -654,8 +644,214 @@ def convert_module(self, script_module, module, module_name, ir_model):
 
         return ir_graph, {}
 
+    def convert_module(self, script_module, module, module_name, ir_model):
```
Comment (J-shang) [inline nni/retiarii/converter/graph_gen.py:647]: why we need both `convert_module` and `_convert_module`?

File: nni/retiarii/converter/graph_gen.py [hunk 2]
```diff
@@ -654,8 +644,214 @@ def convert_module(self, script_module, module, module_name, ir_model):
...
+    and shape info acquired through `torch.jit.trace`.
+
+    Known issues
+    ------------
+    1. `InputChoice` and `ValueChoice` not supported yet.
+    2. Currently random inputs are feeded while tracing layerchoice.
+       If forward path of candidates depends on input data, then wrong path will be traced.
+       This will result in incomplete shape info.
+    """
+    def convert_module(self, script_module, module, module_name, ir_model, example_inputs):
```
Comment (J-shang) [inline nni/retiarii/converter/graph_gen.py:686]: `example_inputs` -> `dummy_input`?

File: nni/retiarii/converter/graph_gen.py [hunk 3]
```diff
@@ -654,8 +644,214 @@ def convert_module(self, script_module, module, module_name, ir_model):
...
+        module.eval()
+
+        ir_graph, attrs = self._convert_module(script_module, module, module_name, ir_model)
+        self.remove_dummy_nodes(ir_model)
+        self._initialize_parameters(ir_model)
+        self._trace_module(module, module_name, ir_model, example_inputs)
+        return ir_graph, attrs
+
+    def _initialize_parameters(self, ir_model: 'Model'):
```
Comment (J-shang) [inline nni/retiarii/converter/graph_gen.py:695]: is this a common use case of type hints?

**Review by liuzhe-lz — APPROVED** [review]

File: docs/en_US/NAS/HardwareAwareNAS.rst [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+==================
+
+.. contents::
+
+EndToEnd Multi-trial SPOS Demo
+------------------------------
+
+Basically, this demo will select the model whose latency satisfy constraints to train.
+
+To run this demo, first install nn-Meter from source code (currently we haven't released this package, so development installation is required).
```
Comment (zheng-ningxin) [inline docs/en_US/NAS/HardwareAwareNAS.rst:11]: Shall we give a link to nn-Meter here?

File: nni/retiarii/converter/graph_gen.py (hunk 1 above)
Comment (ultmaster) [reply nni/retiarii/converter/graph_gen.py:647]: `_convert_module` is exposed to subclass to overwrite internal logic.

File: nni/retiarii/converter/graph_gen.py (hunk 3 above)
Comment (ultmaster) [reply nni/retiarii/converter/graph_gen.py:695]: I think so. It's used "everywhere" in Retiarii.

**Review by J-shang — APPROVED** [review]

File: docs/en_US/NAS/HardwareAwareNAS.rst (hunk 4 above)
Comment (ultmaster) [reply docs/en_US/NAS/HardwareAwareNAS.rst:11]: I'm not sure. My suggestion is that you can do that later when you actually have nn-Meter open sourced.

**Review by zheng-ningxin — APPROVED** [review]

**Pull request merged**